consensus/merkle_proof/src/lib.rs

```diff
@@ -159,16 +181,115 @@ impl MerkleTree {
         matches!(self, MerkleTree::Leaf(_))
     }
 
+    /// Finalize deposits up to deposit with count = deposits_to_finalize
+    pub fn finalize_deposits(
+        &mut self,
+        deposits_to_finalize: usize,
+        level: usize,
+    ) -> Result<(), MerkleTreeError> {
+        match self {
+            MerkleTree::Finalized(_) => Ok(()),
+            MerkleTree::Zero(_) => Err(MerkleTreeError::ZeroNodeFinalized),
+            MerkleTree::Leaf(hash) => {
+                *self = MerkleTree::Finalized(*hash);
+                Ok(())
+            }
+            MerkleTree::Node(hash, left, right) => {
+                let deposits = 0x1 << level;
+                if deposits <= deposits_to_finalize {
+                    *self = MerkleTree::Finalized(*hash);
+                    return Ok(());
+                }
+                left.finalize_deposits(deposits_to_finalize, level - 1)?;
+                let remaining = (deposits_to_finalize as i64) - (deposits as i64) / 2;
+                if remaining > 0 {
+                    right.finalize_deposits(remaining as usize, level - 1)?;
+                }
+                Ok(())
+            }
+        }
+    }
+
+    fn append_finalized_branches(&self, result: &mut Vec<H256>) {
+        match self {
+            MerkleTree::Zero(_) | MerkleTree::Leaf(_) => {}
+            MerkleTree::Finalized(h) => result.push(*h),
+            MerkleTree::Node(_, left, right) => {
+                left.append_finalized_branches(result);
+                right.append_finalized_branches(result);
+            }
+        }
+    }
+
+    pub fn get_finalized_snapshot(&self) -> Vec<H256> {
+        let mut result = vec![];
+        self.append_finalized_branches(&mut result);
+        result
+    }
+
+    pub fn from_finalized_snapshot(
+        finalized_branches: &[H256],
+        deposits: usize,
+        level: usize,
+    ) -> Result<Self, MerkleTreeError> {
+        if finalized_branches == EMPTY_SLICE {
+            return if deposits == 0 {
+                Ok(MerkleTree::Zero(level))
+            } else {
+                Err(MerkleTreeError::InvalidSnapshot(
+                    InvalidSnapshot::EmptyBranchesWithNonZeroDeposits(deposits),
+                ))
+            };
+        }
+        if deposits == (0x1 << level) {
+            return Ok(MerkleTree::Finalized(finalized_branches[0]));
```

Comment:
This function only ever indexes into the first element. So the only way this could ever panic would be if the array was empty. But the empty array case is checked above, so the code could never reach this point.

I would usually try and use safe functions but I can't see a way to do it here without making the code needlessly complicated. There's no default value to fall back on if indexing should fail in this case or the case below if I try to use `split_first`.

Unless you have a suggestion.

Reply:
Okay nevermind [I did find a way to do it](https://gist.github.com/ethDreamer/491408d85f7fb68a13bfda6c6795bd39) but I still think this is kinda more convoluted than the way it is written now. What do you think?

Reply:
My bad, I agree the way it is now is better

Reply:
I actually did find a better way to rewrite this function that is still clear :)
What do you think?